// zenn-markdown-html/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.enablePreview = void 0;
var markdown_it_1 = __importDefault(require("markdown-it"));
var md_line_number_1 = require("./utils/md-line-number");
// plugis
var md_container_1 = require("./utils/md-container");
var md_renderer_fence_1 = require("./utils/md-renderer-fence");
var md_linkify_to_card_1 = require("./utils/md-linkify-to-card");
var md_prism_1 = require("./utils/md-prism");
var md_katex_1 = require("./utils/md-katex");
var md_br_1 = require("./utils/md-br");
var md_custom_block_1 = require("./utils/md-custom-block");
var markdown_it_imsize_1 = __importDefault(require("@steelydylan/markdown-it-imsize"));
var markdown_it_anchor_1 = __importDefault(require("markdown-it-anchor"));
var mdContainer = require('markdown-it-container');
var mdFootnote = require('markdown-it-footnote');
var mdTaskLists = require('markdown-it-task-lists');
var mdInlineComments = require('markdown-it-inline-comments');
var mdLinkAttributes = require('markdown-it-link-attributes');
var md = markdown_it_1.default({
    breaks: true,
    linkify: true,
});
md.linkify.set({ fuzzyLink: false });
md.use(md_br_1.mdBr)
    .use(md_prism_1.mdPrism)
    .use(md_renderer_fence_1.mdRendererFence)
    .use(markdown_it_imsize_1.default)
    .use(md_custom_block_1.mdCustomBlock)
    .use(mdContainer, 'details', md_container_1.containerDetailsOptions)
    .use(mdContainer, 'message', md_container_1.containerMessageOptions)
    .use(mdFootnote)
    .use(mdTaskLists, { enabled: true })
    .use(mdInlineComments)
    .use(mdLinkAttributes, {
    pattern: /^(?!https:\/\/zenn\.dev\/)/,
    attrs: {
        rel: 'nofollow',
    },
})
    .use(markdown_it_anchor_1.default, {
    level: [1, 2, 3, 4],
    permalink: markdown_it_anchor_1.default.permalink.ariaHidden({
        placement: 'before',
        class: 'header-anchor-link',
        symbol: '',
    }),
    tabIndex: false,
})
    .use(md_katex_1.mdKatex)
    .use(md_linkify_to_card_1.mdLinkifyToCard);
// custom footnote => TODO: ファイルを分ける
md.renderer.rules.footnote_block_open = function () {
    return '<section class="footnotes">\n' +
        '<div class="footnotes-title"><img src="https://twemoji.maxcdn.com/2/svg/1f58b.svg" class="emoji footnotes-twemoji" loading="lazy" width="20" height="20">脚注</div>\n' +
        '<ol class="footnotes-list">\n';
};
var markdownToHtml = function (text) {
    if (!(text && text.length))
        return '';
    return md.render(text);
};
exports.default = markdownToHtml;
exports.enablePreview = function () {
    md_line_number_1.mdLineNumber(md);
};
